File: rag/rag_answer.py
"""
RAG Generation — Gemini 2.0 Flash via Vertex AI.
Builds the prompt, calls the model, extracts citations.

Also used by test_citations.py for testing prompt construction.
"""
import re
from config import VERTEX_PROJECT_ID, VERTEX_LOCATION, GENERATION_MODEL


CITATION_RE = re.compile(r"\[Fonte:[^\]]{5,200}\]", re.IGNORECASE)


SYSTEM_PROMPT = """Sei un esperto di EU AI Act (Regolamento UE 2024/1689) e normativa AI.
Rispondi ESCLUSIVAMENTE in base ai documenti forniti nel contesto.
Per ogni affermazione cita la fonte nel formato:
  [Fonte: {titolo_documento}, {sezione_ref}, p. {numero_pagina}]
Se l'informazione non è nel contesto, di' esplicitamente:
  "Questa informazione non è disponibile nei documenti indicizzati."
Non inventare riferimenti normativi. Non citare articoli che non compaiono nel contesto."""


def build_prompt(query: str, context_chunks: list[dict]) -> str:
    """
    Construct the RAG prompt with context chunks.
    Each chunk includes its source citation so the model can reference it.

    Used both by answer_with_rag() and by tests.
    """
    if not context_chunks:
        return (
            f"{SYSTEM_PROMPT}\n\n"
            f"CONTESTO: nessun documento rilevante trovato.\n\n"
            f"Domanda: {query}\n\n"
            f"Risposta: Questa informazione non è disponibile nei documenti indicizzati. "
            f"Non puoi rispondere senza fonti nel contesto."
        )

    context_parts = []
    for i, chunk in enumerate(context_chunks, 1):
        ref = chunk.get("section_ref") or "sezione sconosciuta"
        page = chunk.get("page_number") or "?"
        title = chunk.get("document_title") or chunk.get("document_id") or "documento"
        sim = chunk.get("similarity") or chunk.get("boosted_similarity") or 0
        context_parts.append(
            f"[FONTE {i}] {title} — {ref} (p. {page}) [rilevanza: {sim:.2f}]\n"
            f"{chunk['chunk_text']}\n"
        )

    context_block = "\n---\n".join(context_parts)

    return (
        f"{SYSTEM_PROMPT}\n\n"
        f"CONTESTO ({len(context_chunks)} fonti):\n"
        f"{context_block}\n\n"
        f"Domanda: {query}\n\n"
        f"Risposta (includi citazioni [Fonte:...] per ogni affermazione):"
    )


def parse_answer(raw_answer: str, context_chunks: list[dict]) -> dict:
    """
    Extract structured output from the model's raw answer.
    Returns: {"answer": str, "sources": list, "confidence": str}
    """
    citations = CITATION_RE.findall(raw_answer)

    # Confidence heuristic: based on citation count and avg similarity
    avg_sim = sum(c.get("similarity", 0) for c in context_chunks) / max(len(context_chunks), 1)
    if citations and avg_sim >= 0.7:
        confidence = "HIGH"
    elif citations and avg_sim >= 0.4:
        confidence = "MEDIUM"
    else:
        confidence = "LOW"

    # Build sources list from context (cited ones first)
    sources = []
    for chunk in context_chunks:
        sources.append({
            "document_title": chunk.get("document_title"),
            "section_ref":    chunk.get("section_ref"),
            "page_number":    chunk.get("page_number"),
            "similarity":     round(chunk.get("similarity", 0), 3),
        })

    return {
        "answer":     raw_answer.strip(),
        "sources":    sources,
        "citations":  citations,
        "confidence": confidence,
    }


def answer_with_rag(query: str, context_chunks: list[dict]) -> dict:
    """
    Generate an answer using Gemini 2.0 Flash with retrieved context.

    Returns:
        {
          "answer": str,
          "sources": list of {document_title, section_ref, page_number, similarity},
          "citations": list of [Fonte:...] strings found in answer,
          "confidence": "LOW" | "MEDIUM" | "HIGH",
        }
    """
    try:
        import vertexai
        from vertexai.generative_models import GenerativeModel
        vertexai.init(project=VERTEX_PROJECT_ID, location=VERTEX_LOCATION)
    except ImportError:
        raise ImportError(
            "google-cloud-aiplatform not installed. "
            "Run: pip install google-cloud-aiplatform"
        )

    prompt = build_prompt(query, context_chunks)
    model = GenerativeModel(GENERATION_MODEL)

    response = model.generate_content(
        prompt,
        generation_config={
            "temperature": 0.1,   # low temperature for factual legal Q&A
            "max_output_tokens": 2048,
            "top_p": 0.95,
        },
    )

    raw_answer = response.text if hasattr(response, "text") else str(response)
    return parse_answer(raw_answer, context_chunks)

File: src/lib/rag/rag-answer.ts
// RAG Generation — Gemini 2.0 Flash via Vertex AI
// Builds the RAG prompt with cited sources, calls generation model,
// extracts [Fonte: ...] citations from the answer.

import type { RagChunk, RagAnswer, RagSource } from "./rag-types";
import { generateText, MODEL_NAME } from "./rag-vertex";

const CITATION_RE = /\[Fonte:[^\]]{5,200}\]/gi;

const SYSTEM_PROMPT_IT = `Sei un esperto di EU AI Act (Regolamento UE 2024/1689) e normativa AI.
Rispondi ESCLUSIVAMENTE in base ai documenti forniti nel contesto.

FORMATO OBBLIGATORIO — rispetta SEMPRE queste regole:
1. Scrivi UNA frase introduttiva opzionale (senza citazione).
2. Poi elenca i punti come bullet con "* " all'inizio di ogni riga.
3. Ogni bullet DEVE terminare con la citazione inline:
   [Fonte: {titolo_documento} — {sezione_ref}, p. {numero_pagina}]
4. Non usare mai prosa continua per le affermazioni principali.
5. Se le informazioni sono parziali, cita quello che trovi come bullet e segnala i limiti.
6. Se l'informazione è completamente assente, scrivi un solo bullet:
   * Informazione non disponibile nei documenti indicizzati. [Fonte: nessun documento pertinente]
Non inventare riferimenti normativi.`;

const SYSTEM_PROMPT_EN = `You are an expert on the EU AI Act (Regulation EU 2024/1689) and AI regulation.
Answer EXCLUSIVELY based on the documents provided in the context.

MANDATORY FORMAT — always follow these rules:
1. Write ONE optional introductory sentence (no citation).
2. Then list points as bullets starting with "* " on each line.
3. Every bullet MUST end with an inline citation:
   [Source: {document_title} — {section_ref}, p. {page_number}]
4. Never use continuous prose for main statements.
5. If information is partial, cite what you find as bullets and note the limits.
6. If information is completely absent, write a single bullet:
   * Information not available in the indexed documents. [Source: no relevant document]
Do not invent regulatory references.`;

function buildPrompt(
  query: string,
  chunks: RagChunk[],
  lang: "it" | "en" = "it"
): string {
  const systemPrompt = lang === "it" ? SYSTEM_PROMPT_IT : SYSTEM_PROMPT_EN;

  if (chunks.length === 0) {
    return (
      `${systemPrompt}\n\n` +
      `CONTESTO: nessun documento rilevante trovato.\n\n` +
      `Domanda: ${query}\n\n` +
      `Risposta: Questa informazione non è disponibile nei documenti indicizzati.`
    );
  }

  const contextParts = chunks.map((c, i) => {
    const ref   = c.sectionRef ?? "sezione sconosciuta";
    const page  = c.pageNumber ?? "?";
    const title = c.documentTitle;
    const sim   = c.similarity.toFixed(2);
    return (
      `[FONTE ${i + 1}] ${title} — ${ref} (p. ${page}) [rilevanza: ${sim}]\n` +
      `${c.chunkText}`
    );
  });

  const contextBlock = contextParts.join("\n\n---\n\n");
  const instruction  = lang === "it"
    ? "Risposta (includi citazioni [Fonte:...] per ogni affermazione):"
    : "Answer (include [Source:...] citations for each statement):";

  return (
    `${systemPrompt}\n\n` +
    `CONTESTO (${chunks.length} fonti):\n` +
    `${contextBlock}\n\n` +
    `Domanda: ${query}\n\n` +
    `${instruction}`
  );
}

function parseCitations(text: string): string[] {
  return text.match(CITATION_RE) ?? [];
}

function computeConfidence(
  citations: string[],
  chunks: RagChunk[]
): "LOW" | "MEDIUM" | "HIGH" {
  const avgSim =
    chunks.reduce((s, c) => s + c.similarity, 0) / Math.max(chunks.length, 1);
  if (citations.length > 0 && avgSim >= 0.7) return "HIGH";
  if (citations.length > 0 && avgSim >= 0.4) return "MEDIUM";
  return "LOW";
}

/**
 * Generate a cited answer using Gemini 2.0 Flash with retrieved context.
 */
export async function answerWithRAG(
  query: string,
  chunks: RagChunk[],
  lang: "it" | "en" = "it"
): Promise<RagAnswer> {
  const t0 = Date.now();
  const prompt = buildPrompt(query, chunks, lang);

  let rawAnswer: string;
  try {
    rawAnswer = await generateText(prompt, {
      temperature:     0.1,
      maxOutputTokens: 2048,
    });
  } catch (err) {
    // Graceful degradation: return error message as answer
    rawAnswer =
      lang === "it"
        ? "Errore nella generazione della risposta. Riprova più tardi."
        : "Answer generation failed. Please try again.";
    console.error("[RAG] Generation error:", err);
  }

  const citations = parseCitations(rawAnswer);
  const confidence = computeConfidence(citations, chunks);

  const sources: RagSource[] = chunks.map((c) => ({
    documentTitle: c.documentTitle,
    sectionRef:    c.sectionRef,
    pageNumber:    c.pageNumber,
    similarity:    Math.round(c.similarity * 1000) / 1000,
  }));

  return {
    answer:     rawAnswer.trim(),
    sources,
    citations,
    confidence,
    model:      MODEL_NAME,
    latencyMs:  Date.now() - t0,
  };
}
